File: services/payment-service-python/app/messaging/consumer.py
"""
RabbitMQ consumer for order events.
"""
import json
import logging
import asyncio
import aio_pika
from typing import Optional

from app.core.config import settings
from app.db import get_db_context
from app.services import PaymentService
from app.messaging.publisher import publish_payment_event
from app.models.payment import PaymentStatus

logger = logging.getLogger(__name__)

_connection: Optional[aio_pika.RobustConnection] = None
_consumer_task: Optional[asyncio.Task] = None


async def process_order_created(message: aio_pika.IncomingMessage) -> None:
    """
    Process an order.created event.
    
    Creates a payment attempt and publishes the result.
    """
    async with message.process():
        try:
            body = json.loads(message.body.decode())
            logger.info(f"Received order.created event: {body}")
            
            # Extract order details
            order_id = body.get("orderId") or body.get("order_id")
            amount = body.get("totalAmount") or body.get("total") or body.get("amount") or 0.0
            
            if not order_id:
                logger.error("Missing order_id in event payload")
                return
            
            # Create payment in database
            with get_db_context() as db:
                payment = PaymentService.create_payment_from_order_event(
                    db=db,
                    order_id=order_id,
                    amount=float(amount),
                )
                
                # Build and publish result event
                event = PaymentService.build_payment_event(payment)
                routing_key = (
                    "payment.success"
                    if payment.status == PaymentStatus.SUCCESS
                    else "payment.failed"
                )
                
                await publish_payment_event(event.model_dump(), routing_key)
                
                logger.info(
                    f"Processed order.created -> payment {payment.id} "
                    f"with status {payment.status.value}"
                )
                
        except json.JSONDecodeError as e:
            logger.error(f"Failed to decode message: {e}")
        except Exception as e:
            logger.error(f"Error processing order.created event: {e}")
            raise


async def start_consumer() -> None:
    """Start consuming order.created events from RabbitMQ."""
    global _connection
    
    try:
        logger.info(f"Starting order event consumer...")
        
        # Connect to RabbitMQ
        _connection = await aio_pika.connect_robust(
            settings.RABBITMQ_URL,
            timeout=30,
        )
        
        channel = await _connection.channel()
        await channel.set_qos(prefetch_count=10)
        
        # Declare order exchange
        order_exchange = await channel.declare_exchange(
            settings.ORDER_EXCHANGE,
            aio_pika.ExchangeType.TOPIC,
            durable=True,
        )
        
        # Declare queue for order.created events
        queue = await channel.declare_queue(
            settings.ORDER_CREATED_QUEUE,
            durable=True,
        )
        
        # Bind queue to exchange
        await queue.bind(
            order_exchange,
            routing_key=settings.ORDER_CREATED_ROUTING_KEY,
        )
        
        logger.info(
            f"Bound queue {settings.ORDER_CREATED_QUEUE} to "
            f"{settings.ORDER_EXCHANGE}/{settings.ORDER_CREATED_ROUTING_KEY}"
        )
        
        # Start consuming
        await queue.consume(process_order_created)
        
        logger.info("Order event consumer started successfully")
        
        # Keep the consumer running
        try:
            await asyncio.Future()
        except asyncio.CancelledError:
            logger.info("Consumer task cancelled")
            
    except Exception as e:
        logger.error(f"Consumer error: {e}")
        raise


async def start_consumer_task() -> None:
    """Start the consumer as a background task."""
    global _consumer_task
    
    _consumer_task = asyncio.create_task(start_consumer())
    logger.info("Consumer background task created")


async def stop_consumer() -> None:
    """Stop the consumer and close connections."""
    global _connection, _consumer_task
    
    if _consumer_task:
        _consumer_task.cancel()
        try:
            await _consumer_task
        except asyncio.CancelledError:
            pass
        _consumer_task = None
        logger.info("Consumer task stopped")
    
    if _connection and not _connection.is_closed:
        await _connection.close()
        _connection = None
        logger.info("Consumer connection closed")

File: services/payment-service/src/events/consumer.ts
import { ConsumeMessage } from 'amqplib';
import { getChannel, EXCHANGE_NAME } from '../config/rabbitmq';
import { PaymentService } from '../services/payment.service';
import { OrderCreatedEvent } from '../types/payment.types';
import logger from '../utils/logger';

const QUEUE_NAME = 'payment-service.order-events';
const ORDER_CREATED_KEY = 'order.created';

export const startEventConsumer = async (): Promise<void> => {
  try {
    const channel = getChannel();

    // Declare queue for this service
    await channel.assertQueue(QUEUE_NAME, {
      durable: true,
    });

    // Bind queue to exchange with routing key
    await channel.bindQueue(QUEUE_NAME, EXCHANGE_NAME, ORDER_CREATED_KEY);

    logger.info(`✅ Queue '${QUEUE_NAME}' bound to exchange '${EXCHANGE_NAME}'`);
    logger.info(`   - Listening for: ${ORDER_CREATED_KEY}`);

    // Set prefetch to process one message at a time
    await channel.prefetch(1);

    // Start consuming
    await channel.consume(QUEUE_NAME, async (msg: ConsumeMessage | null) => {
      if (!msg) {
        return;
      }

      const routingKey = msg.fields.routingKey;

      try {
        const content = msg.content.toString();
        const event = JSON.parse(content);

        logger.info(`📥 Received event: ${routingKey}`, {
          eventId: event.eventId,
          orderId: event.orderId,
        });

        if (routingKey === ORDER_CREATED_KEY) {
          await handleOrderCreated(event as OrderCreatedEvent);
        } else {
          logger.warn(`Unknown routing key: ${routingKey}`);
        }

        // Acknowledge message
        channel.ack(msg);
        logger.debug(`Message acknowledged: ${routingKey}`);
      } catch (error) {
        logger.error(`Error processing message: ${routingKey}`, error);

        // Negative acknowledge - requeue if first attempt, otherwise discard
        const requeue = !msg.fields.redelivered;
        channel.nack(msg, false, requeue);

        if (!requeue) {
          logger.warn(`Message discarded after retry: ${routingKey}`);
        }
      }
    });

    logger.info('✅ Payment event consumer started successfully');
  } catch (error) {
    logger.error('Failed to start event consumer:', error);
    throw error;
  }
};

/**
 * Handle order.created event
 * Process payment for the order
 */
const handleOrderCreated = async (event: OrderCreatedEvent): Promise<void> => {
  logger.info(`Processing order.created for order: ${event.orderId}`);

  try {
    await PaymentService.processPayment(event);
    logger.info(`Payment processing completed for order: ${event.orderId}`);
  } catch (error) {
    logger.error(`Failed to process payment for order: ${event.orderId}`, error);
    throw error;
  }
};
